tests/index.js

```diff
@@ -50,79 +74,131 @@ describe('broccoli-sane-watcher', function (done) {
 
     var builder = new broccoli.Builder(filter);
     watcher = new Watcher(builder);
-    watcher.on('change', function (results) {
-      assert.equal(results.directory, 'output');
+    watcher.on('change', eventHandler(function (results) {
+      expect(results.directory).to.equal('output');
+
       if (changes++) {
-        done();
+        defer.resolve();
       } else {
         fs.writeFileSync('tests/fixtures/a/file.js');
       }
-    });
+    }, defer));
+
+    watcher.on('error', errorHandler(defer));
+
+    return defer.promise;
+  });
+
+  it('cleans up heimdall nodes after builds and rebuilds', function() {
+    function countNodes() {
+      var count = 0;
 
-    watcher.on('error', function (error) {
-      assert.ok(false, error.message);
-      done();
+      heimdall.visitPreOrder(function () {
+        ++count;
+      });
+
+      return count;
+    }
+
+    var defer = RSVP.defer();
+
+    expect(countNodes()).to.equal(1, 'there is only the heimdall root node');
+
+    fs.mkdirSync('tests/fixtures/a');
+    var filter = new TestFilter(['tests/fixtures/a'], function () {
+      return 'output';
     });
+
+    var builder = new broccoli.Builder(filter);
+    watcher = new Watcher(builder);
+
+    watcher.on('error', errorHandler(defer));
+
+    var builds = 0;
+
+    watcher.on('change', eventHandler(function () {
+      // we build 3 times, but each time we only have the nodes from the current
```

Comment:
@stefanpenner that is exactly what [eventHandler](https://github.com/ember-cli/broccoli-sane-watcher/blob/f3794626c009a615e268a1971ab46d550602f8d4/tests/index.js#L33-L41) is intended to do